### PR TITLE
Parameter fragments on Push and Pull implemented and parameter handling improved

### LINKED ISSUE #258: Revit_Toolkit: Parameter Mapping for Push

### LINKED ISSUE #391: Take out ToBHoMFramingElementProperty from ToBHoMFramingElement
### Description:
As in the header, would be good to extract the `ToBHoMFramingElement` method from the inside of `BH.UI.Revit.Engine.Convert.ToBHoMFramingElement`. However, the fact that the framing property is dependent on instance parameter (_BHE_Material Grade_) needs to be captured somehow.

### LINKED ISSUE #392: Merge ToRevitFamilySymbol_Column and ToRevitFamilySymbol_Framing into one method
#### Broken rules:
As in the header, methods sitting in `BH.UI.Revit.Engine.Convert`.

### LINKED ISSUE #395: Framing justification/offset bugs
#### Description:
Beam justification is not captured correctly in some cases (e.g. pushing a created from scratch framing element with justification or offset, pull -> push of twisted, rotated beams etc.).

#### How to replicate:
Try to pull -> push beams from Beams.rvt on [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FPhysical)

#### Test file(s):
As above

### LINKED ISSUE #578: Revit_Toolkit: SetCustomData should not copy the parameters that are directly related to object properties
#### Description:
Atm `SetCustomData` by default copies all parameters to `CustomData`, resulting e.g. in Level having `Elevation` property and `Elevation` key in `CustomData`, which is confusing and needs to be sorted out

### LINKED ISSUE #622: Revit_Toolkit: Element z-justification
#### Description:
This might be a well known bug/behaviour, but I couldn't find any issues about it.

The location of IFramingElement/Bar after being pulled does not correspond properly with their location in Revit.

Specificly z-justification
but also noticed that end extension was not used,

Now these thing are placed on the custom data after the pull, but it's not very useful unless you're pushing back to Revit again(?)

This is probably a feature request, but you'll know best @pawelbaran 

#### Test file(s):

https://burohappold.sharepoint.com/:f:/s/BHoM/EvT8YPLsFNtMu0GhOSQqZFMB69dmZcVvQll1v_3iXA0S7A?e=ultinw

### LINKED ISSUE #723: Revit_Toolkit: Handling of parameters vs properties on Push/Pull
### Detailed issue:

Currently CRUD in Revit_Toolkit works in the following way:
- on **Read** the Revit element is converted into a BHoM object, which implies loss of some Revit-specific information (e.g. a beam with driving curve aligned to its top face will be pulled as a Physical beam with driving curve aligned to its centroid etc.)
- after Revit element => BHoM object conversion on **Read**, all Revit parameters are copied over to the `CustomData` of a BHoM object, resulting in duplication of some information (e.g. pulled `Level` will have `Elevation` property and Elevation key in `CustomData`)
- on **Create** the Revit element is first created based on a BHoM object and then the `CustomData` is copied over to its parameters, resulting in a risk of changing the originally created element
- on **Update** location of the Revit element is first updated based on a BHoM object and then the `CustomData` is copied over to its parameters, which can easily conflict with each other

To make it more complex, very often a single BHoM property is not 1:1 derived from Revit parameters, e.g.:
- location of a beam in Revit is defined by a combination of its driving curve and multiple parameters (attachment to other elements, offsets, justification etc.), in BHoM it is a single curve
- material of an element is defined by a combination of _Material_ parameter and _Material Grade_ parameter
- in language version other than English, `Elevation` of a BHoM `Level` would be 1:1 correspondent with parameter under different name (_Rzędna_ in 🇵🇱)

Combination of the above creates a multitude of possible (apparent) issues on Pull/Push:
- pulling a beam, moving its driving curve up/down and pushing back will result in no change because of it being overconstrained by the previously pulled parameters being pushed back
- pulling a level, using `SetProperty` on its `Elevation` and pushing back to Revit will result in no change because first the Revit element will be updated by the property, but then changed back by the previously pulled elevation parameter that stayed unchanged
- (...)

### TL;DR:
The current approach to pulling and pushing Revit parameters does not work because always all of them get pulled and pushed, causing various interferences with BHoM object properties.

### Proposed solution:
- on **Read** the location of BHoM object would be calculated to ensure that it represents the Revit element (e.g. BHoM representation of a Revit beam aligned to its top face would be moved by half of its height downwards due to the fact that BHoM beam does not carry offsets)
- later on **Read** all parameters would be stored in `RevitParameters` fragment
- the parameters to be overwritten on **Push** would be stored in BHoM object's `CustomData` (therefore would need to be explicitly added  by the user before pushing)
- on **Create** the location of a Revit element would be calculated to ensure that it represents the BHoM object, taken into account the modifications of Revit element imposed by parameters pushed from `CustomData` (inverse of first point)
- on **Update** the location of a Revit element would be calculated to ensure that it represents the BHoM object, taken into account the modifications of Revit element imposed by parameters pushed from combination of `CustomData` and parameters of the element being updated (same as above)

This brings us to the following workflow for the end user:

image

The above will require the user to make an explicit request to change a chosen parameter, while still exposing all parameters on Pull (very useful) and leaving the Push formula fully open (anything can changed, conflicts explained at the beginning can be imposed by the end user if she/he likes).

I am rather against using a Fragment for pushed parameters because it will be yet another step to take for the user. Simple `SetProperty` will be way more intuitive I believe.

### Other considered solutions:
- disallowing certain parameters from being overwritten - a straight road down to hell, identifying the ones and making sure they are well maintained would be a disaster + opennes and hackability would be limited (against BHoM principle)
- adding a switch to `RevitPushConfig` to push or not push parameters - this would introduce an _everything or nothing_ approach that would not resolve the primary issue of _some_ parameters being mishandled (still _all_ of them would be pushed, including the troublemakers)
- adding an enum on `RevitPushConfig` to decide whether a BHoM object or its `CustomData` is meant to push - above issue with _everything or nothing_ still valid + disastrous UX (_why is it not pushing parameters any more?_)

Please feel free if anything above is unclear or needs more visual explanation. Happy to push this one a bit as I would like to give it a first shot next week @al-fisher @rwemay 

Thanks!

### LINKED ISSUE #727: Revit_Toolkit: Remove AddAdjacentSpaceId and AddSpaceId
#### Description:
Both unused and  deprecated - not exposed to the end user so can be deleted straight away.

### LINKED ISSUE #740: Revit_Toolkit: Framing pulls at the centerline, but pushes back in at the justification point
#### Description:
I'm attempting to update section properties for beams via GH. So, I pulled revit FramingElement.Beam, changed the Property, and the pushed them back with PushType UpdateOnly. The beams each dropped half their depth.

#### Steps to reproduce:
pull Beams, then push them back in using UpdateOnly. They will be lowered.

#### Expected behaviour:
What comes in should match what goes out.

### LINKED ISSUE #742: Revit_Toolkit: Element types do not get updated based on updated BHoM object's properties
#### Description:

Following @JosefTaylor: in order to update type in Revit, atm one needs to `SetProperty` _Type_ to the ElementId of the chosen family type. It would be more handy to be able to update the type based on BHoM object's property, even based on currently implemented name matching mechanism.

### PR DESCRIPTION
**Pull request by pawelbaran**
#### THIS PR WILL BREAK THE SCRIPTS BASED ON LINKING REVIT PARAMETERS VIA CUSTOMDATA - maintaining parameters in both `CustomData` and `IRevitParameters` was virtually impossible plus that shift must have happened at some point anyways, so I took the sooner the better approach. 
The user will need to choose between the two:
- update the script a bit
- stick to 3.1

### Issues addressed by this PR

Closes #258
Closes #391
Closes #392
Closes #395
Closes #578
Closes #622
Closes #723
Closes #727
Closes #740
Closes #742

### Test files
- Most test have been put together in [one folder](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue723%2DParametersVsProperties&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4):
    - _GeneralTests_ covers principles (pull, push, parameter mapping) and is meant to explain how do the things work now
    - _Framing_ focuses on framing (curved, twisted beams fail and that is not going to be covered by this PR)
    - _Columns_ focuses on columns - to be extended further

The above are meant to become standard test files and samples after a bit of review and refinement. Running a few [macro tests](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4) is also highly recommended.

Besides that, problem specific file for #391 is available [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue391%2DPullFramingProperty&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).

### Changelog

Following (mainly) #723:
- 3 `IParameterFragments` have been implemented:
    - `RevitPulledParameters` - immutable, to store parameters that have been pulled from Revit
    - `RevitIdentifiers` - identifiers of a pulled Revit element, also immutable
    - `RevitParametersToPush` - a fragment containing the parameters that are meant to be overwritten on push - thanks to that, the user has control what will be updated and what not
- `SetLocationOnUpdate` property has been added to `RevitPushConfig` in order to allow updating parameters only (without forcing location update)
- The approach to updating location has been changed in a way that ensures that a pushed element is placed in same location as its correspondent BHoM object
- Parameter mapping on push has been implemented
- Update of types based on BHoM object property (`Property` for framing, `Construction` for panels etc.) has been enabled
- Necessary support methods have been added to unlock the above
- Code base has been further cleaned up

### Additional comments

I am aware that this PR is gargantuan and dropped on the very last day - apologies for that, but figuring out all Revit conventions and edge cases is a laborious process that took slightly longer than expected (however, looks like it finally works!). Since it would be good to get this PR reviewed and merged soon, **I am happy to accept minor issues found during the review process - I will do my best to get them fixed next week**.

**Please do not be scared by the number of changed files** many of them are just renaming of methods/properties 🤣

### CODE REVIEW

**Comment by pawelbaran** [comment]
@FraserGreenroyd it looks like AppVeyor has an issue with Sockets and Adapter - not sure why, could you please have a look? Thanks!

**Comment by kThorsager** [comment]
### In the test: `Framing`
while pulling the Original, non twisted objects. Not a problem for the others.
And first updating then use the Create only, I get this:
image
Which seems strange, but might just be Revit magic.

I also see that the CustomData is empty, and as the pulling and pushing is working I guess that's fine, but I'm unsure on if it's required to have the adapterId in the CustomData to align with other adapters. But unsure.

### In the test: `GeneralTests` 
you're pushing some elements with properties named: `Non-existantProperty`. I can't find it in the model afterwards, and there's no error/warning. feels like there should be one of those at least.

### The Structure Pull macro shows:
Pulling walls is still problematic, where some holes are missing (might be strange planarity), some are on the location, and some are as openings.
Pulling Structural Panels is also a problem, no vertical panels are pulled and one gets this message:
image
But testing on a project model does pulls the vertical panels as well so 🤷‍♂️

**Comment by pawelbaran** [comment]
Thank you @kThorsager for the review, I've replied to your comments below.

> ### In the test: `Framing`
> Which seems strange, but might just be Revit magic.

Indeed, it looks like Revit is going mad when framing elements are 1. duplicated 2. close to each other - it is often the case for Revit that it tried to be smarter than you 😃 You can try `DeleteThenCreate` to make sure that there is no interference between the existing and newly created elements.
 
> I also see that the CustomData is empty, and as the pulling and pushing is working I guess that's fine, but I'm unsure on if it's required to have the adapterId in the CustomData to align with other adapters. But unsure.

Yes, `CustomData` is meant to be empty after the changes made in this PR. `AdapterId` is a good question though - @alelom, is it necessary to store it in the `CustomData`? For me it is a one-liner.

> ### In the test: `GeneralTests`
> you're pushing some elements with properties named: `Non-existantProperty`. I can't find it in the model afterwards, and there's no error/warning. feels like there should be one of those at least.

`NonExistentProperty` does not exist and therefore nothing happens - I was testing what will happen if the user sets wrong (nonexistent) property name.

> ### The Structure Pull macro shows:
> Pulling walls is still problematic, where some holes are missing (might be strange planarity), some are on the location, and some are as openings.
> Pulling Structural Panels is also a problem, no vertical panels are pulled and one gets this message:

Part of the code base responsible for panel converts has not been touched in this PR and there is a range of issues raised (#429, #445, #620, #709, #710) that I am planning to take on as the next big task.

> But testing on a project model does pulls the vertical panels, so 🤷‍♂️

😀

**Comment by IsakNaslundBh** [comment]
@pawelbaran to problem with the CI is most likely coming from the fact that you have this branch https://github.com/BHoM/BHoM/tree/Revit_Toolkit-%23723-ParametersVsProperties pushed to the BHoM, which is now generally behind master.

If required, that branch needs to be PRed and noted as dependant, otherwise, should close it, after which @FraserGreenroyd can retrigger the CI.

**Comment by pawelbaran** [comment]
Good spot @IsakNaslundBh. Both BHoM and BHoM_Engine branches are not needed any more, will close now.

**Comment by IsakNaslundBh** [comment]
@FraserGreenroyd actually found it, to give praise to the right person 😄

**Comment by pawelbaran** [comment]
Manual rebase done, should be possible to merge now.
